pkg/cmd/app.go

```diff
@@ -236,180 +252,188 @@ func newOSWatcher(sigs ...os.Signal) chan os.Signal {
 }
 
 func action(c *cli.Context) (err error) {
-restart:
-
-	// The purpose of this function is to capture any panic that may occur
-	// during initialization and return an error.
-	defer func() {
-		if r := recover(); r != nil {
-			logrus.WithField(dcgmexporter.LoggerStackTrace, string(debug.Stack())).Error("Encountered a failure.")
-			err = fmt.Errorf("encountered a failure; err: %v", r)
-		}
-	}()
-
-	logrus.Info("Starting dcgm-exporter")
-	config, err := contextToConfig(c)
-	if err != nil {
-		return err
-	}
-
-	if config.Debug {
-		// enable debug logging
-		logrus.SetLevel(logrus.DebugLevel)
-		logrus.Debug("Debug output is enabled")
-	}
-
-	logrus.Debugf("Command line: %s", strings.Join(os.Args, " "))
-
-	logrus.WithField(dcgmexporter.LoggerDumpKey, fmt.Sprintf("%+v", config)).Debug("Loaded configuration")
+	ctx, cancel := context.WithCancel(context.Background())
+	return stdout.Capture(ctx, func() error {
```

Comment:
@rohit-arora-dev , What is the HoF?

Reply:
Higher Order Functions

Reply:
The HoS doesn't have any impact on testability. However, implementation without HoS will significantly increase the complexity of the capturer.

Reply:
Unfortunately, the proposed approach with Start/Stop doesn't work well for the use case when we capture GO and C stdout because of the global nature of the stdout.  

The GO runtime and C runtime have different implementations of the `STDOUT` and `IO` implementations.  The `os.Stdout` is not the same file that C uses when interacting with `syscall.Stdout`. When we create a pipe to read and forward `syscall.Stdout`, we get the `syscall.Stdout` file descriptor and replace it with a file descriptor of the pipe writer. After this, ALL writes to `/dev/stdout` inside the process go to the pipe and processing. If we run the code as part of the unit test, the unit test framework output will also go the pipe. That is why restoring the process's original state is critical. The implemented functional approach ensure that changes in `syscall.Stdout` doesn't escape of the function implementation. Also, the usage of `syscall` is unsafe code, meaning it is runs outside of the GO runtime and GO garbage collector cleans the stack after function execution. 

All these things make it difficult to implement the suggested Start/Stop approach, as it has side effects.

Reply:
Thanks for giving it a try.